### DISCUSSION TITLE
Introducing MartenDB support into OrchardCore module data layers

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [OrchardCMS/OrchardCore#11184] [Ideas]
### MartenDB document databases and/or event sources
Core technologies include `postgresql`, `JSON`, or rather also `BSON` to be more precise, and `MartenDB`.

### Solution
Are there modules already? _NuGet_ search did not reveal anything, so I am assuming probably not. I'm not sure I'd want to replace the `Orchard` data layers, but might like to at least _contain_ if you will those business features, `MartenDB`, etc, via a module. Which I gather is the strength and draw of an `Orchard`, correct?

### Versus "traditional" ORM style approaches
Or versus a _RavenDB_, has some compelling features I'd like to leverage for some of our more _business_ back end and middle tier oriented features.

**Comment by hishamco** [comment]
Hi @mwpowellhtx I already checked MartenDb few months ago, you could use it for your custom modules if you want. While the OrchardCore are using YesSQL which allows you to use SQLite, SQL Server, MySQL and PostgreSQL as well. JSON is used undearneath to store the actual data. For more information you could check [YesSQL](https://github.com/sebastienros/yessql) repo.

One more thing OC is already did a good work to separate the data layer, so you can find:

- OC.Data.Abstractions
- OC.Data
- OC.Data.YesSql.Abstractions
- OC.Data.YesSql

So nothing prevent you to use what you make you happy

**Comment by mwpowellhtx** [comment]
Interesting; how come y'all rolled your own `NoSql`?

**Reply by hishamco** [reply]
Seems you need to do something similar in terms of Data Layer separation, I already did some progress in Orchard Core Contrib (OCC) to enable EF Core as data layer

**Reply by mwpowellhtx** [reply]
Well, need/want, but yeah. If we are going to adopt this approach, we kind of like what MartenDB has to offer for document database purposes. Also the compelling event sourcing.

**Reply by hishamco** [reply]
Frankly I didn't dig into MartenDB but you got the idea if you want to use another data layer in case YesSQL not fit for you

**Reply by mwpowellhtx** [reply]
To be clear, I'm not sure we want to replace the entire data layer. Only focused on our module(s). That said, appreciate all the feedback.

**Reply by hishamco** [reply]
For that, I advice you to work with YesSQL unless there's something cruicial letting you to use MartenDB or any other data layer
